IS212-G9-T5/skill-based-role-portal
# feat: view applicants T5-68&22

## 📋 Overview (what is this PR about)

This PR mainly addresses the critical user journey of viewing applicants who have applied for the role listings. ([T5-22](https://spmg9t5.atlassian.net/browse/T5-22)) Note that this feature is only available for `HR`,`Admin`, and `Manager`.

I have also added route protections for each page to ensure only authorised user has access to certain features of the system ([T5-68](https://spmg9t5.atlassian.net/browse/T5-68))

## 🚀 What are the features/changes included?
- [x] Backend endpoint to return all role listings without unneeded fields (matched/unmatched skills)
- [x] Landing page for user whose role has the right to view applicants for open role listings
> - This page is protected (role-based authorisation)
> - Navbar tailored for viewing applications and creating applications
> - A table that shows all the role listings (including both `open` and `closed` roles)
> - Table is paginated and sortable
> - Search bar that allows searching feature using role name
> - Edit icon which will redirect user to `update-role-listing` page
> - Profile icon which will redirect user to `view-applicants` page
- [x] Page to view all applicants for a particular role listing
> - Role listing card component at the top of the page to shows all details of the role listing
> - A list of applicants who have applied to this role
> - Applicant card is clickable. Once clicked, full details of the applicant will be shown at the right side of the page
- [x] Responsiveness for both pages

## 🧪 What testing have I done to make sure it is working?

### Frontend
```
yarn install
yarn build
yarn dev
```
Able to access frontend w/o any errors

### Backend
```
poetry shell
poetry install
poetry run python -m flask --app wsgi:app run
```
Able to access endpoints w/o any errors

## 🧐 How should the reviewer go about testing it?
1. Login as different users and complete all necessary user journeys required for this project
2. Ensure all features above work as intended

## ✅ Final check
- [x] I have done the following:
  - `<git rebase main>` to replay your feature commits from the main branch
  - If there are any conflicts, resolve them then run:
    - `git add <resolved rebase conflict files>`
    - `<git rebase --continue>`
    - You may need to run `<git pull>` if your local feature branch is behind the remote feature branch
    - `<git push>`

## 🖼️ Make it visual ✨ (if possible, attach a screenshot of what the reviewer is supposed to see)

https://github.com/IS212-G9-T5/skill-based-role-portal/assets/68802506/c1493b84-9d35-4fb4-8074-30ccc101caee

 
## ✍️ Any additional notes

Thank you reviewer! Sorry for the large PR >< Let me know if there is anything that needs to be changed / improved on!


[T5-22]: https://spmg9t5.atlassian.net/browse/T5-22?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[T5-68]: https://spmg9t5.atlassian.net/browse/T5-68?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
